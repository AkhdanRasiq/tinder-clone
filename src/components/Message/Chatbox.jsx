import React from 'react'

class Chatbox extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      strTest: ""
    }
  }

  render() {
    return(
      <div>
        
      </div>
    )
  }
}

export default Chatbox
