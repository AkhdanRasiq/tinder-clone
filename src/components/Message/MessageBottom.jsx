import React from 'react'

import TextField from '@mui/material/TextField'
import SendIcon from '@mui/icons-material/Send'
import { IconButton } from "@mui/material"

import { onSendWebsocket } from '../../utils/adapters'
import { handleGenerateMessage } from '../../utils/util'

class MessageBottom extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      strMessage: "",
      arrMessages: []
    }
  }

  componentDidUpdate = (prevProps) => {
    if(this.props.a_arrMessages !== prevProps.a_arrMessages)
      this.setState({ arrMessages: this.props.a_arrMessages })
  }

  handleMessageToSend = () => {
    const generatedMessage = handleGenerateMessage(this.state.strMessage)
    onSendWebsocket(generatedMessage);

    this.setState({ arrMessages: [...this.state.arrMessages, generatedMessage]})
    this.props.a_messageOnUpdateCallback([...this.state.arrMessages, generatedMessage])

    this.setState({ strMessage: "" })
  }

  handleSendMessage = (event) => {
    if (event.key === 'Enter' && this.state.strMessage !== '')
      this.handleMessageToSend()
  }

  handleBtnSendMessage = (event) => {
    event.preventDefault()
    if (this.state.strMessage !== '')
      this.handleMessageToSend()
  }

  handleTyping = (event) => {
    this.setState({ strMessage: event.target.value });
  }

  render() {
    return (
      <div className    ='messageBottomContainer'>

        <TextField
          fullWidth
          className     ='messageBottom_textfield'
          placeholder   ='Type a message'
          autoComplete  ='off'
          onKeyDown     ={this.handleSendMessage} 
          onChange      ={this.handleTyping}
          value         ={this.state.strMessage}  
        />

        <IconButton onClick={this.handleBtnSendMessage}>
          <SendIcon
            className   ='messageBottom_btnSend'
            fontSize    ='large'
          />
        </IconButton>
      </div>
    )
  }
}

export default MessageBottom
