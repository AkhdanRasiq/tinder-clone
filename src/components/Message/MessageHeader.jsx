import React from 'react'

import { Avatar } from '@mui/material'
import { styled } from '@mui/material/styles'
import Badge from '@mui/material/Badge'

const StyledBadgeOnline = styled(Badge)(({ theme }) => ({
  '& .MuiBadge-badge': {
    backgroundColor: '#44b700',
    color: '#44b700',
    width: '10px',
    height: '10px',
    borderRadius: '100%',
    right: 10,
    bottom: 15,
    boxShadow: `0 0 0 2px ${theme.palette.background.paper}`,
    '&::after': {
      position: 'absolute',
      top: -1,
      left: -1,
      width: '100%',
      height: '100%',
      borderRadius: '100%',
      animation: 'ripple 1.2s infinite ease-in-out',
      border: '1px solid currentColor',
      content: '""',
    },
  },
  '@keyframes ripple': {
    '0%': {
      transform: 'scale(.8)',
      opacity: 1,
    },
    '100%': {
      transform: 'scale(2.4)',
      opacity: 0,
    },
  },
}));

const StyledBadgeOffline = styled(Badge)(({ theme }) => ({
  '& .MuiBadge-badge': {
    backgroundColor: '#b70000',
    color: '#b70000',
    width: '10px',
    height: '10px',
    borderRadius: '100%',
    right: 10,
    bottom: 15,
    boxShadow: `0 0 0 2px ${theme.palette.background.paper}`
  }
}));

class MessageHeader extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      isOnline: false
    }
  }

  componentDidMount() {
    window.addEventListener("authorOnline", this.handleEventStatus);
  }

  componentWillUnmount() {
    window.removeEventListener("authorOnline", this.handleEventStatus);
  }

  handleEventStatus = (event) => {
    this.setState({ isOnline: event.detail.isOnline })
  }

  render() {
    return (
      <div className='messageHeaderContainer'>
        {this.state.isOnline ?
          <StyledBadgeOnline
            overlap="circular"
            anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
            variant="dot"
          >
            <Avatar className='messageHeaderAvatar' alt="Avatar Image" src="https://www.logo.wine/a/logo/Tinder_(app)/Tinder_(app)-Flame-Logo.wine.svg" />
          </StyledBadgeOnline>
        :
          <StyledBadgeOffline
            overlap="circular"
            anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
            variant="dot"
          >
            <Avatar className='messageHeaderAvatar' alt="Avatar Image" src="https://www.logo.wine/a/logo/Tinder_(app)/Tinder_(app)-Flame-Logo.wine.svg" />
          </StyledBadgeOffline>
        }
      </div>
    )
  }
}

export default MessageHeader
