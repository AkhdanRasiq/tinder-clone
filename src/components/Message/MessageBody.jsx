import React from 'react'
import Chatbox from './Chatbox'

class MessageBody extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      arrMessages: []
    }
  }

  componentDidUpdate = (prevProps) => {
    if(this.props.a_arrMessages !== prevProps.a_arrMessages)
      this.setState({ arrMessages: this.props.a_arrMessages })
  }

  render() {
    return (
      <div className='messageBodyContainer' id='messageBodyContainer'>
        {this.state.arrMessages.map((msg) => (
          <Chatbox 
            key={msg.timestamp}
            a_messages={msg}
          />
        ))}
      </div>
    )
  }
}

export default MessageBody
