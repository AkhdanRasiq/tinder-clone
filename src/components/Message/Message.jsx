import React, { Component } from 'react'
import { connectWebsocket } from '../../utils/adapters'

import MessageHeader from '../Message/MessageHeader'
import MessageBody from '../Message/MessageBody'
import MessageBottom from '../Message/MessageBottom'

class Message extends Component {
  constructor(props) {
    super(props)
    this.state = {
      arrMessages: []
    }
  }

  componentDidMount = () => {
    connectWebsocket()
    window.addEventListener("newMessage", this.handleEventNewMessage);
  }

  componentWillUnmount() {
    window.removeEventListener("newMessage", this.handleEventNewMessage);
  }

  handleEventNewMessage = (event) => {
    this.setState({ arrMessages: [...this.state.arrMessages, event.detail] })
  }

  handleMessagesUpdate = (event) => {
    this.setState({ arrMessages: event })
  }

  render() {
    return (
      <div className="messageContainer">
        <MessageHeader />
        <MessageBody
          a_arrMessages             = {this.state.arrMessages}
        />

        <MessageBottom
          a_arrMessages             = {this.state.arrMessages}
          a_messageOnUpdateCallback = {this.handleMessagesUpdate}
        />
      </div>
    )
  }
}

export default Message
