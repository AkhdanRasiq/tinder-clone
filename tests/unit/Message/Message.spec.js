import React from 'react'
import { shallow, mount } from "enzyme"
import { SocketIO, Server } from 'mock-socket'

// custom imports
import Message from "../../../src/components/Message/Message"


describe('Message.jsx', () => {
  let wrapper, shallowWrapper
  const mockServer = new Server('ws://test')
  window.io = SocketIO

  beforeEach(async () => {
    const mockUpObject = {
      focus: () => null,
    };
    global.document.getElementById = jest.fn(() => mockUpObject);

    wrapper = mount(<Message />)
    mockServer.on('connection', socket => {
      socket.emit('task', 'test message from mock server')
    })
    shallowWrapper = shallow(<Message />)
  })

  it("renders properly", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("has Message component", () => {
    expect(shallowWrapper.find('.messageContainer').length).toBe(1)
  })
})
