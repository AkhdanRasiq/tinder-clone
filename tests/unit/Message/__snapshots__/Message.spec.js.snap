// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message.jsx renders properly 1`] = `ReactWrapper {}`;
